storage/range_tree.go

```diff
@@ -18,31 +18,306 @@
+type (
+	// RangeTree represents holds the metadata about the RangeTree.  There is
+	// only one for the whole cluster.
+	rangeTree proto.RangeTree
+	// RangeTreeNode represents a node in the RangeTree, each range
+	// has only a single node.
+	rangeTreeNode proto.RangeTreeNode
+)
+
+// TODO(bram): Add a cache
+// TODO(bram): Add parent keys
+// TODO(bram): Add delete for merging ranges.
+// TODO(bram): Add more elaborate testing.
+
+// set saves the RangeTree node to the db.
+func (n *rangeTreeNode) set(db *client.KV) error {
+	if err := db.PutProto(engine.RangeTreeNodeKey(n.Key), (*proto.RangeTreeNode)(n)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// set saves the RangeTreeRoot to the db.
+func (t *rangeTree) set(db *client.KV) error {
+	if err := db.PutProto(engine.KeyRangeTreeRoot, (*proto.RangeTree)(t)); err != nil {
+		return err
+	}
+	return nil
+}
 
+// SetupRangeTree creates a new RangeTree.  This should only be called as part of
+// store.BootstrapRange.
+func SetupRangeTree(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp, startKey proto.Key) error {
 	tree := &proto.RangeTree{
 		RootKey: startKey,
 	}
-
 	node := &proto.RangeTreeNode{
 		Key:   startKey,
 		Black: true,
 	}
-
+	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, tree); err != nil {
+		return err
+	}
 	if err := engine.MVCCPutProto(batch, ms, engine.RangeTreeNodeKey(startKey), timestamp, nil, node); err != nil {
 		return err
 	}
+	return nil
+}
 
-	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, tree); err != nil {
-		return err
+// GetRangeTree fetches the RangeTree proto.
+func getRangeTree(db *client.KV) (*rangeTree, error) {
+	tree := &proto.RangeTree{}
+	ok, _, err := db.GetProto(engine.KeyRangeTreeRoot, tree)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree:%s", engine.KeyRangeTreeRoot)
+	}
+	return (*rangeTree)(tree), nil
+}
+
+// getRangeTreeNode return the RangeTree node for the given key.  If the key is
+// nil, an empty RangeTreeNode is returned.
+func getRangeTreeNode(db *client.KV, key *proto.Key) (*rangeTreeNode, error) {
+	if key == nil {
+		return nil, nil
+	}
+	node := &proto.RangeTreeNode{}
+	ok, _, err := db.GetProto(engine.RangeTreeNodeKey(*key), node)
+	if err != nil {
+		return nil, err
 	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree node:%s",
+			engine.RangeTreeNodeKey(*key))
+	}
+	return (*rangeTreeNode)(node), nil
+}
 
+// getRootNode returns the RangeTree node for the root of the RangeTree.
+func (t *rangeTree) getRootNode(db *client.KV) (*rangeTreeNode, error) {
+	return getRangeTreeNode(db, &t.RootKey)
+}
+
+// InsertRange adds a new range to the RangeTree.  This should only be called
+// from operations that create new ranges, such as range.splitTrigger.
+func InsertRange(db *client.KV, key proto.Key) error {
+	tree, err := getRangeTree(db)
```

Comment:
I've been thinking about this.  We'll be able to insert as a right child of the range being split.  But I'm not convinced that this will actually save any reads.  From that point, we will have to walk down until we hit a leaf, then walk right back up the whole tree rotating/flipping where needed.  With the cache being added in the next PR, it will minimize the reads, even when starting from the root.  I'll drop a todo to revisit it.